### PR TITLE
Submit widget query form before saving edit widget modal.

### LINKED ISSUE #7922: Do not require search execution inside dashboard widget edit modal to save widget search controls changes.
When a user changes the search controls (query, streams, timerange) of a dashboard widget, he needs to execute the search inside the widget edit modal, before clicking on the modal "Save" button, to save the changes.
The behaviour can be confusing for some users and we should save all changes independently of the search execution.

This is a follow-up improvement for https://github.com/Graylog2/graylog2-server/pull/7829 which changed the searchbar execution behaviour.

* Graylog Version: current development version for 3.3

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

Before this change, a successful change to the timerange/stream selection/query of a widget on a dashboard required submitting the widget's search after the change and before clicking "Save". This was unintuitive and unnecessarily complicated.

This change is now submitting the widget's query form when the user clicks "Save" and the form is dirty, not requiring the user to have submitted the form manually before.

Fixes #7922.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by dennisoelkers** [comment]
One issue that this is producing is:

```
react-dom.development.js:88 Warning: validateDOMNesting(...): <form> cannot appear as a descendant of <form>.
    in form (created by EventListConfiguration)
```

The form element is now including all of the modal. Therefore no further forms can be used in the modal's body.

**Comment by dennisoelkers** [comment]
I "solved" the nested form issue by removing the form element around the events checkbox. This is a viable solution for now, but we would need to address this if we want to use forms in the edit widget modal (it might make sense to restrict it though).

File: graylog2-web-interface/src/views/components/widgets/EditWidgetFrame.jsx [hunk 1]
```diff
@@ -40,51 +44,69 @@ type Props = {
...
+           dialogComponentClass={EditWidgetDialog}
+           enforceFocus={false}>
+      <SearchBarForm initialValues={{ timerange, streams, queryString }}
+                     onSubmit={_onSubmit}>
+        <div className={styles.gridContainer}>
+          <IfDashboard>
+            <Modal.Header className={styles.QueryControls}>
+              <QueryEditModeContext.Provider value="widget">
+                <HeaderElements />
+                <WidgetQueryControls widget={widget} />
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/widgets/EditWidgetFrame.jsx:88]: `widget={widget}` can be removed

File: graylog2-web-interface/src/views/components/WidgetQueryControls.jsx [hunk 2]
```diff
@@ -11,29 +10,22 @@ import DocumentationLink from 'components/support/DocumentationLink';
...
 import QueryInput from './searchbar/AsyncQueryInput';
-import SearchBarForm from './searchbar/SearchBarForm';
-
-import { DEFAULT_TIMERANGE } from '../Constants';
 
 type Props = {
   availableStreams: Array<any>,
   config: any,
   globalOverride: ?GlobalOverride,
-  widget: Widget,
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/WidgetQueryControls.jsx:36]: The widget prop can also be removed in `WidgetQueryControls.test.jsx`

File: graylog2-web-interface/src/views/components/widgets/SaveOrCancelButtons.jsx [hunk 3]
```diff
@@ -8,11 +10,22 @@ type Props = {
...
-);
+const SaveOrCancelButtons = ({ onFinish, onCancel }: Props) => {
+  const { handleSubmit, dirty } = useFormikContext();
+  const _onFinish = useCallback((...args) => {
+    if (handleSubmit && dirty) {
+      handleSubmit();
+    }
+
+    return onFinish(...args);
+  }, [onFinish, handleSubmit]);
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/widgets/SaveOrCancelButtons.jsx:21]: The current implementation does not work as expected (the behaviour on save did not changed), looks the reason is that `dirty` is missing in this array.

File: graylog2-web-interface/src/views/components/WidgetQueryControls.test.jsx [hunk 4]
```diff
@@ -58,11 +56,19 @@ describe('WidgetQueryControls', () => {
   const emptyGlobalOverride = {};
   const globalOverrideWithQuery = { query: { type: 'elasticsearch', query_string: 'source:foo' } };
 
-  const renderSUT = (props = {}) => render(
+  const Wrapper = ({ children }: { children: React.Node }) => (
     <WrappingContainer>
+      <SearchBarForm initialValues={{ timerange: { type: 'relative', range: 300 }, queryString: '', streams: [] }} onSubmit={() => {}}>
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/WidgetQueryControls.test.jsx:61]: You can use `DEFAULT_TIMERANGE` from `views/Contsants` here.

**Review by linuspahl — CHANGES_REQUESTED** [review]
After I figured out why the on save behaviour did not changed, everything worked like a charm!

File: graylog2-web-interface/src/views/components/widgets/EditWidgetFrame.jsx (hunk 1 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/widgets/EditWidgetFrame.jsx:88]: :white_check_mark:

File: graylog2-web-interface/src/views/components/WidgetQueryControls.jsx (hunk 2 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/WidgetQueryControls.jsx:36]: :white_check_mark:

File: graylog2-web-interface/src/views/components/WidgetQueryControls.test.jsx (hunk 4 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/WidgetQueryControls.test.jsx:61]: :white_check_mark:

File: graylog2-web-interface/src/views/components/widgets/SaveOrCancelButtons.jsx (hunk 3 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/widgets/SaveOrCancelButtons.jsx:21]: :white_check_mark:

**Comment by dennisoelkers** [comment]
Thanks for the great review, @linuspahl!

**Review by linuspahl — APPROVED** [review]
Looks good to me!

**Pull request merged**